### DISCUSSION TITLE
Testing elements on specific inputs

### DISCUSSION THREAD

**Discussion by SethPoulsen** [PrairieLearn/PrairieLearn#7530] [Dev]
Inspired by this and similar past issues: https://github.com/PrairieLearn/PrairieLearn/pull/7527#issuecomment-1514893625

Do we want to have a way to test specific _inputs_ for our element testing? 

Its not clear what the format would look like, since different elements take inputs in different formats. I supposed we could just have a system where you manually write the JSON that should be serialized to the server for the specific input you want to test.

**Comment by eliotwrobson** [comment]
Having a system to manually write JSON is probably the most robust. However, elements like the `pl-big-o-input`, `pl-symbolic-input`, and `pl-units-input` all use an architecture that allows testing the grading algorithm using `pytest`, which covers a similar use case to this the proposed discussion here. It allows for manually inputting a suite of specific inputs to the grading algorithm, and there should be an easy way to extend this to fuzz testing. I think an architecture like this could handle the use case detailed in the linked comment with substantially lower developer overhead than extending the existing PL test suite to handle JSON files.

**Reply by jonatanschroeder** [reply]
It seems to me that `pytest` would be mostly for tests the backend. Would it support a bit on the frontend (in particular a whole sequence of parse and grade)?

**Reply by SethPoulsen** [reply]
Ah yes, that's a good point. I would like to point out that `pl-order-blocks` does also have extensive unit testing of the grading algorithm using Pytest (edit: in fact, it had this long before any of those other elements did!). The issue is, I suppose, that in `pl-order-blocks`, there is a lot of stuff going on in between the user interface, and the grading algorithm, and that's where things have gone badly. 

Maybe the difference between `pl-order-blocks` and the others is that `pl-order-blocks` has a more complicated input mechanism and configuration options, and so the code to convert from the input mechanism to the format for the grading algorithm is more complicated. 

Fuzz testing on the level of the grading algorithm would have prevented some, but not all of the bugs we have seen with `pl-order-blocks`.

**Reply by SethPoulsen** [reply]
@jonatanschroeder replied at the same time as me so I didn't see his, I think we are both pointing out the same thing, that it is hard to test the whole sequence of parsing and grading with the current set up.

**Reply by nwalters512** [reply]
Keeping as much as possible in `pytest` will go a long way towards minimizing runtime for our test suite. Our integration tests are quite slow, and they'll get even slower if we try to do true E2E tests with a headless browser. We could add one or two tests for interactivity + frontend JS, but IMO we shouldn't use that for exhaustively testing grading logic.

If we want to test the full parse + grade flow, that should be completely feasible to do in pure Python: just construct a `data` object with raw submissions matching the shape we'd get from the client, then call `parse()` and `grade()`!

**Reply by SethPoulsen** [reply]
@nwalters512 ya, that's basically exactly what I meant when I said 

> manually write the JSON that should be serialized to the server

I guess I could do just for `pl-order-blocks` using pytest without having to worry about setting up some system that works for all the other elements though.

**Reply by SethPoulsen** [reply]
It would be nice if it integrated somehow with the `test` function that the elements defines though.

**Reply by nwalters512** [reply]
If you want to use `test()`, just call it! I don't think we need anything to "integrate", though there might be some useful abstractions that fall out of this work. We can worry about that later.

**Reply by eliotwrobson** [reply]
Per the discussion above, it's doable (but a bit more effort) to test a full `parse` + `grade` flow purely in `pytest`. However, even without that additional effort, the current `grade` function for `pl-order-blocks` isn't fully covered by the existing `pytest` tests (just some of the grading algorithms). This is what I meant about changing the existing architecture. To have better test coverage, there should be better abstraction between the grading code itself and the `grade` function. In the other elements I mentioned, those grading algorithms are abstracted heavily from the structure of the element itself, making them very easy to test.

> If you want to use `test()`, just call it! I don't think we need anything to "integrate", though there might be some useful abstractions that fall out of this work. We can worry about that later.

Unfortunately, anything within an element controller is not easily importable from another Python file without some additional effort because of the naming convention. However per the rest of my comment, I don't think this is really super necessary.

**Reply by nwalters512** [reply]
I'm also in favor of abstractions where appropriate, but there's nothing stopping us from just changing the Python element controller file names to use underscores instead of dashes if that's a roadblock.

**Reply by eliotwrobson** [reply]
> I'm also in favor of abstractions where appropriate, but there's nothing stopping us from just changing the Python element controller file names to use underscores instead of dashes if that's a roadblock.

I think that's a fine course of action, but if this is done it should accompany some work on a larger testing framework that prevents code duplication for all tests of this type across elements (at the very least, come up with something that's extensible enough that it could be applied to other elements later, since this will effectively be adding a third testing architecture to PL). That's why I mentioned the existing level of abstraction used for `pytest` in the other elements, it requires pretty minimal effort to set up with the current architecture and catches _most_ of the common issues.
